src/xcode/ENA/ENA/Source/Workers/Keychain/KeychainHelper.swift

```diff
@@ -19,51 +19,70 @@
 
 import Foundation
 
-enum KeychainHelper {
-	// swiftlint:disable:next force_unwrapping
-	private static let _service = Bundle.main.bundleIdentifier!
+enum KeychainError: Error {
+	case initializationError
+	case clearItem(reason: String? = nil)
+	case save(reason: String? = nil)
+	case keyGenerationFail
+}
 
-	static func saveToKeychain(key: String, data: Data) -> Bool {
+struct KeychainHelper {
 
+	/// The service this helper is registered to
+	let service: String
+
+	/// Initializer
+	/// - Parameter service: the service to use; defaults to the main bundle id
+	// swiftlint:disable:next force_unwrapping
+	init(service: String = Bundle.main.bundleIdentifier!) throws {
+		self.service = service
+		if self.service.isEmpty { throw KeychainError.initializationError }
+	}
+
+	func clearInKeychain(key: String) throws {
 		let deleteResult = SecItemDelete(
 			.keychainQueryForDeleting(
 				account: key,
-				service: _service
+				service: service
 			)
 		)
-
+		// ignore 'item not found errors' as this might happen, e.g. on first launch
 		if deleteResult != errSecSuccess && deleteResult != errSecItemNotFound {
-			if let message = SecCopyErrorMessageString(deleteResult, nil) {
-				logError(message: "Failed to delete keychain item '\(key)' due to: \(message as String)")
-			} else {
-				logError(message: "Failed to delete keychain item '\(key)' due to unknown error")
-			}
+			let message = SecCopyErrorMessageString(deleteResult, nil) ?? "unknown error" as CFString
+			let reason = "Failed to delete existing keychain item '\(key)' due to \(message)"
+			throw KeychainError.clearItem(reason: reason)
 		}
+	}
 
+	@discardableResult
+	func saveToKeychain(key: String, data: Data) throws -> Bool {
+		try clearInKeychain(key: key)
 		let addResult = SecItemAdd(
 			.keychainQueryForAdding(
 				account: key,
-				service: _service,
+				service: service,
 				data: data
 			),
 			nil
 		)
 
 		if addResult != errSecSuccess {
 			if let message = SecCopyErrorMessageString(addResult, nil) {
-				logError(message: "Failed to add keychain item '\(key)' due to: \(message as String)")
+				let message = "Failed to add keychain item '\(key)' due to: \(message as String)"
+				throw KeychainError.save(reason: message)
 			} else {
-				logError(message: "Failed to add keychain item '\(key)' due to unknown error")
+				let message = "Failed to add keychain item '\(key)' due to unknown error"
+				throw KeychainError.save(reason: message)
 			}
 		}
 
-		return addResult == errSecSuccess
+		return true
```

Comment:
If the functions `throws` and always returns true, why not remove the return type?

Reply:
I had the same thoughts but just wanted to finish and don't start a refactoring spree removing all these oldschool C-style patterns. There are tons of such issues, especially with the SQL handling. This is definitely something for my own refactoring list.

Reply:
So you didn't refactor it because it would be to big of an undertaking? Would be huge if you only refactored this occurrence?

Reply:
It seems I removed all reference to the old return value. This is refactored out and should be reflected here soon.